File: bindings/tbdex_uniffi/src/messages/order_status.rs
use crate::errors::{Result, TbdexError};
use futures::executor::block_on;
use std::sync::{Arc, RwLock};
use tbdex::{
    json::{FromJson, ToJson},
    messages::order_status::{OrderStatus as InnerOrderStatus, OrderStatusData},
};
use web5_uniffi_wrapper::dids::bearer_did::BearerDid;

pub struct OrderStatus(pub Arc<RwLock<InnerOrderStatus>>);

impl OrderStatus {
    pub fn create(
        to: String,
        from: String,
        exchange_id: String,
        data: OrderStatusData,
        protocol: Option<String>,
        external_id: Option<String>,
    ) -> Result<Self> {
        let order_status =
            InnerOrderStatus::create(&to, &from, &exchange_id, &data, protocol, external_id)?;

        Ok(Self(Arc::new(RwLock::new(order_status))))
    }

    pub fn sign(&self, bearer_did: Arc<BearerDid>) -> Result<()> {
        let mut inner_order_status = self.0.write().map_err(TbdexError::from_poison_error)?;
        inner_order_status.sign(&bearer_did.0.clone())?;
        Ok(())
    }

    pub fn from_json_string(json: &str) -> Result<Self> {
        let inner_order_status = InnerOrderStatus::from_json_string(json)?;

        Ok(Self(Arc::new(RwLock::new(inner_order_status))))
    }

    pub fn to_json_string(&self) -> Result<String> {
        let inner_order_status = self.0.read().map_err(TbdexError::from_poison_error)?;

        Ok(inner_order_status.to_json_string()?)
    }

    pub fn get_data(&self) -> Result<InnerOrderStatus> {
        let order_status = self.0.read().map_err(TbdexError::from_poison_error)?;

        Ok(order_status.clone())
    }

    pub fn verify(&self) -> Result<()> {
        let order_status = self.0.read().map_err(TbdexError::from_poison_error)?;

        Ok(block_on(order_status.verify())?)
    }
}

File: bound/typescript/src/messages/order-status.ts
import { MessageMetadata } from ".";
import { BearerDid } from "../dids/bearer-did";
import { tbdexError } from "../errors";
import wasm from "../wasm";

export class OrderStatus {
  readonly metadata: MessageMetadata;
  readonly data: OrderStatusData;
  signature: string;

  constructor(
    metadata: MessageMetadata,
    data: OrderStatusData,
    signature: string
  ) {
    this.metadata = metadata;
    this.data = data;
    this.signature = signature;
  }

  static fromJSONString = (json: string): OrderStatus => {
    const obj = JSON.parse(json);
    return new OrderStatus(obj.metadata, obj.data, obj.signature);
  };

  toJSONString = (): string => {
    return JSON.stringify({
      metadata: this.metadata,
      data: this.data,
      signature: this.signature,
    });
  };

  static create = (
    to: string,
    from: string,
    exchangeId: string,
    data: OrderStatusData,
    protocol?: string,
    externalId?: string
  ): OrderStatus => {
    try {
      let json = wasm.order_status_create(
        to,
        from,
        exchangeId,
        JSON.stringify(data),
        protocol,
        externalId
      );
      let obj = JSON.parse(json);
      return new OrderStatus(obj.metadata, obj.data, obj.signature);
    } catch (error) {
      throw tbdexError(error);
    }
  };

  sign = (bearerDid: BearerDid) => {
    try {
      const signature = wasm.order_status_sign(
        this.toJSONString(),
        bearerDid.toWASM()
      );
      this.signature = signature;
    } catch (error) {
      throw tbdexError(error);
    }
  };

  verify = async () => {
    try {
      await wasm.order_status_verify(this.toJSONString());
    } catch (error) {
      throw tbdexError(error);
    }
  };
}

export type OrderStatusData = {
  details?: string;
  status: string;
};
